sdk/core/azure-core/src/datetime.cpp

```diff
@@ -4,869 +4,832 @@
 {
-  return CumulativeDaysToMonth[month] + monthDay + (IsLeapYear(year) && month > 1) - 1;
+  throw std::invalid_argument(std::string("Error parsing DateTime: ") + description + ".");
 }
 
-constexpr int CountLeapYears(int yearsSince1601)
+template <typename T>
+T ParseNumber(
+    std::string::size_type* cursor,
+    std::string const& str,
+    std::string::size_type strLen,
+    char const* description,
+    int minLength,
+    int maxLength)
 {
-  int tmpYears = yearsSince1601;
-
-  int const year400 = tmpYears / 400;
-  tmpYears -= year400 * 400;
+  if (*cursor + minLength <= strLen)
+  {
+    auto const MaxChars = std::min(static_cast<int>(strLen - *cursor), maxLength);
+    int64_t value = 0;
+    auto i = 0;
+    for (; i < MaxChars; ++i)
+    {
+      int const ch = str[*cursor + i];
+      if (std::isdigit(ch))
+      {
+        value = (value * 10) + (static_cast<decltype(value)>(ch) - '0');
+        continue;
+      }
 
-  int result = year400 * 97;
+      if (i < minLength)
+      {
+        value = -1;
+      }
 
-  int const year100 = tmpYears / 100;
-  tmpYears -= year100 * 100;
-  result += year100 * 24;
+      break;
+    }
 
-  result += tmpYears / 4;
+    if (value >= 0 && value <= std::numeric_limits<T>::max())
+    {
+      *cursor += i;
+      return static_cast<T>(value);
+    }
+  }
 
-  return result;
+  ThrowParseError(description);
+  return T();
 }
 
-constexpr int64_t AdjustTimezone(
-    int64_t result,
-    unsigned char chSign,
-    int adjustHours,
-    int adjustMinutes)
+template <typename T>
+void IncreaseAndCheckMinLength(
+    std::string::size_type* minLength,
+    std::string::size_type actualLength,
+    T increment)
 {
-  if (adjustHours > 23)
+  *minLength += static_cast<std::string::size_type>(increment);
+  if (actualLength < *minLength)
   {
-    return -1;
+    ThrowParseError("too short");
   }
+}
 
-  // adjustMinutes > 59 is impossible due to digit 5 check
-  int const tzAdjust = adjustMinutes * 60 + adjustHours * 60 * 60;
-  if (chSign == '-')
+void ParseSingleChar(
+    std::string::size_type* cursor,
+    std::string const& str,
+    char const* parsingArea,
+    char ch)
+{
+  if (str[*cursor] != ch)
   {
-    if (std::numeric_limits<int64_t>::max() - result < tzAdjust)
-    {
-      return -1;
-    }
-
-    result += tzAdjust;
+    ThrowParseError(parsingArea);
   }
-  else
-  {
-    if (tzAdjust > result)
-    {
-      return -1;
-    }
+  ++*cursor;
+}
 
-    result -= tzAdjust;
+void ParseSingleOptionalChar(
+    std::string::size_type* cursor,
+    std::string const& str,
+    char ch,
+    std::string::size_type* minLength,
+    std::string::size_type actualLength)
+{
+  if (str[*cursor] == ch)
+  {
+    ++*cursor;
+    IncreaseAndCheckMinLength(minLength, actualLength, 1);
   }
-
-  return result;
 }
 } // namespace
 
-/*
-https://tools.ietf.org/html/rfc822
-https://tools.ietf.org/html/rfc1123
+DateTime const DateTime::SystemClockEpoch = GetSystemClockEpoch();
```

Comment:
I don't quite understand what the purpose of this is. What does `GetSystemClockEpoch` do and why is it needed?

Reply:
It is initialization for static class member.
System clock's zero could be 1970, or 1601. Standard does not specify. Which is why GetSystemClockEpoch() finds it out, and returns it in terms of Azure::Core::DateTime, where it is certainly not zero.
Static private class member SystemClockEpoch is needed so that conversion from `std::chrono::system_clock::time_point` works, and is inlineable:
```
DateTime(std::chrono::system_clock::time_point const& systemTime)
        : DateTime(
            SystemClockEpoch + std::chrono::duration_cast<duration>(systemTime.time_since_epoch()))
    {
    }
```